dm/relay/binlog_writer.go

```diff
@@ -64,6 +74,53 @@ func NewBinlogWriter(logger log.Logger, relayDir string) *BinlogWriter {
 	}
 }
 
+// run starts the binlog writer.
+func (w *BinlogWriter) run() {
+	var (
+		buf       = &bytes.Buffer{}
+		errOccurs bool
+	)
+
+	// writeToFile writes buffer to file
+	writeToFile := func() {
+		if buf.Len() == 0 {
+			return
+		}
+
+		if w.file == nil {
+			w.err.CompareAndSwap(nil, terror.ErrRelayWriterNotOpened.Delegate(errors.New("file not opened")))
+			errOccurs = true
+			return
+		}
+		n, err := w.file.Write(buf.Bytes())
+		if err != nil {
+			w.err.CompareAndSwap(nil, terror.ErrBinlogWriterWriteDataLen.Delegate(err, n))
```

Comment:
sorry seems the `nil` for CAS is not what I expected

```
func TestCAS(t *testing.T) {
	var nilErr error
	e := atomic.NewError(nil)
	e.Store(nilErr)
	require.NoError(t, e.Load())
	ok := e.CompareAndSwap(nilErr, errors.New("test"))
	require.True(t, ok)
}
```

Can we use the `nilErr` variable or simply `Store`? but in the latter we will store the last error not the first one.